File: apps/android/protocol/src/main/kotlin/com/relayium/protocol/RelayRenewProbe.kt
package com.relayium.protocol

/**
 * The 59-byte data-path control frame that proves a migration actually moved
 * the media path — `relay-renew-v1.md` section 6.1.
 *
 * ```
 * [0x0d][ver=1][type][epoch u32BE][round u32BE][nonce 16B][tag 32B]
 * ```
 *
 * ## Why a dedicated kind and not the text codec
 *
 * `0x0d` is outside every kind already in use — file lane 1–8, text 9,
 * pre-upload 12 and the lifecycle bytes — and `relayium-link-v1.md` section 7.3
 * already requires a text-lane frame whose first byte is not `0x09` to be
 * SILENTLY IGNORED. That is what makes this frame safe to send to a peer that
 * has never heard of renewal.
 *
 * Routing a handshake through ordinary kind-9 content would instead be a hard
 * failure on all three clients whenever the conversation is not open — this
 * module's own [TextLaneSession] answers `CONTENT_BEFORE_ACTIVATION` — and it
 * would consume the strictly increasing AEAD `seq`. This frame consumes no
 * sequence and changes no file or text semantics.
 *
 * ## What the tag does and does not prove
 *
 * The HMAC authenticates key possession and freshness. It does NOT authenticate
 * the route. The path evidence is local observation of the selected candidate's
 * generation, plus the peer's own observation, which is the only thing that
 * makes it send an ack. Nothing in this file may be read as route proof.
 */
object RelayRenewProbe {

    /** The registered first byte. Outside every kind in use. */
    const val KIND = 13

    const val VERSION = 1
    const val TYPE_PROBE = 1
    const val TYPE_ACK = 2

    const val NONCE_BYTES = 16
    const val TAG_BYTES = 32

    /** The EXACT length. A frame of any other length is a reject. */
    const val FRAME_BYTES = 2 + 1 + 4 + 4 + NONCE_BYTES + TAG_BYTES

    private const val OFFSET_TYPE = 2
    private const val OFFSET_EPOCH = 3
    private const val OFFSET_ROUND = 7
    private const val OFFSET_NONCE = 11
    private const val OFFSET_TAG = OFFSET_NONCE + NONCE_BYTES

    /**
     * Whether this text-lane frame belongs to the renewal demux, by FIRST BYTE
     * alone.
     *
     * Deliberately not "is a valid probe". The demux has to consume every frame
     * that claims this kind, including a malformed one: passing a 58-byte
     * `0x0d` frame through to the text session would spend one of its inbound
     * rate-budget tokens and reset the ten-minute idle clock, which is exactly
     * what section 6.2 forbids. Validity is [decode]'s job, and its answer is
     * "drop", not "forward".
     */
    fun isControlFrame(frame: ByteArray): Boolean =
        frame.isNotEmpty() && (frame[0].toInt() and 0xff) == KIND

    /** One decoded control frame. The tag is NOT verified here. */
    data class Frame(
        val type: Int,
        val epoch: Long,
        val round: Long,
        val nonce: ByteArray,
        val tag: ByteArray,
    ) {
        val isProbe: Boolean get() = type == TYPE_PROBE

        override fun equals(other: Any?) = other is Frame &&
            other.type == type && other.epoch == epoch && other.round == round &&
            other.nonce.contentEquals(nonce) && other.tag.contentEquals(tag)

        override fun hashCode(): Int {
            var result = type
            result = 31 * result + epoch.hashCode()
            result = 31 * result + round.hashCode()
            result = 31 * result + nonce.contentHashCode()
            result = 31 * result + tag.contentHashCode()
            return result
        }
    }

    /**
     * Read a control frame, or null.
     *
     * Cheap bounds first, exactly in the order section 6.4 states: length, kind
     * byte, version, type. Nothing here allocates on a frame it will refuse
     * beyond the two field copies, and nothing here spends an HMAC.
     */
    fun decode(frame: ByteArray): Frame? {
        if (frame.size != FRAME_BYTES) return null
        if ((frame[0].toInt() and 0xff) != KIND) return null
        if ((frame[1].toInt() and 0xff) != VERSION) return null
        val type = frame[OFFSET_TYPE].toInt() and 0xff
        if (type != TYPE_PROBE && type != TYPE_ACK) return null
        return Frame(
            type = type,
            epoch = Bytes.readUInt32(frame, OFFSET_EPOCH),
            round = Bytes.readUInt32(frame, OFFSET_ROUND),
            nonce = frame.copyOfRange(OFFSET_NONCE, OFFSET_NONCE + NONCE_BYTES),
            tag = frame.copyOfRange(OFFSET_TAG, FRAME_BYTES),
        )
    }

    /** Build one control frame. */
    fun encode(type: Int, epoch: Long, round: Long, nonce: ByteArray, tag: ByteArray): ByteArray {
        require(type == TYPE_PROBE || type == TYPE_ACK) { "type is probe or ack" }
        require(RelayRenewWire.isUint32(epoch) && RelayRenewWire.isUint32(round)) {
            "epoch and round are uint32"
        }
        require(nonce.size == NONCE_BYTES) { "the nonce is $NONCE_BYTES bytes" }
        require(tag.size == TAG_BYTES) { "the tag is $TAG_BYTES raw bytes" }
        val out = ByteArray(FRAME_BYTES)
        out[0] = KIND.toByte()
        out[1] = VERSION.toByte()
        out[OFFSET_TYPE] = type.toByte()
        Bytes.writeUInt32(out, OFFSET_EPOCH, epoch)
        Bytes.writeUInt32(out, OFFSET_ROUND, round)
        nonce.copyInto(out, OFFSET_NONCE)
        tag.copyInto(out, OFFSET_TAG)
        return out
    }

    // ── signing, and why it lives here ──────────────────────────────────────
    //
    // The tag rides the frame RAW while this module's only HMAC primitive
    // speaks standard padded base64, and the raw/encoded conversion belongs
    // beside the frame rather than in every caller. It also keeps the byte
    // helpers — which are internal to this module — out of the app layer, so
    // the engine that drives a `PeerConnection` handles no key material and no
    // encoding of its own.

    /** A freshly signed probe or ack frame, or null if the key is unusable. */
    fun sign(
        keys: Crypto.SessionKeys,
        type: Int,
        from: String,
        to: String,
        epoch: Long,
        round: Long,
        nonce: ByteArray,
    ): ByteArray? {
        if (nonce.size != NONCE_BYTES) return null
        val tag = Bytes.unbase64OrNull(
            Crypto.signAuth(keys, payload(type, from, to, epoch, round, Bytes.base64(nonce))),
        ) ?: return null
        return encode(type, epoch, round, nonce, tag)
    }

    /**
     * Whether a decoded frame's tag is this link's, for the direction stated.
     *
     * [from] is the SENDER and [to] this client, so a relay reflecting a probe
     * back at its sender verifies the reversed tuple and fails — the same
     * property a link leave's payload has, and for the same reason.
     */
    fun verify(keys: Crypto.SessionKeys, frame: Frame, from: String, to: String): Boolean {
        val payload = payload(
            frame.type, from, to, frame.epoch, frame.round, Bytes.base64(frame.nonce),
        )
        return Crypto.verifyAuth(keys, payload, Bytes.base64(frame.tag))
    }

    /** A nonce's canonical key, for the bounded already-acked map. */
    fun nonceKey(nonce: ByteArray): String = Bytes.base64(nonce)

    /** Constant-time nonce comparison, for matching an ack to this side's own. */
    fun nonceEquals(a: ByteArray, b: ByteArray): Boolean = Bytes.constantTimeEquals(a, b)

    /**
     * The EXACT bytes a probe or ack tag covers.
     *
     * The domain-separating `kind` is what makes a probe and its ack different
     * strings, so an observed probe cannot be replayed back as its own ack; the
     * `from`/`to` tuple closes reflection for the same reason a leave's does;
     * and the per-attempt random nonce closes replay across attempts. The nonce
     * is carried as STANDARD PADDED base64, the same encoding the rest of this
     * protocol's tags and keys use.
     */
    fun payload(
        type: Int,
        from: String,
        to: String,
        epoch: Long,
        round: Long,
        nonceBase64: String,
    ): String {
        require(type == TYPE_PROBE || type == TYPE_ACK) { "type is probe or ack" }
        return Json.stringify(
            Json.obj(
                "kind" to Json.of(if (type == TYPE_PROBE) "link-renew-probe" else "link-renew-ack"),
                "from" to Json.of(from),
                "to" to Json.of(to),
                "epoch" to Json.of(epoch),
                "round" to Json.of(round),
                "nonce" to Json.of(nonceBase64),
            ),
        )
    }
}

File: apps/RelayiumKit/Sources/RelayiumKit/Realtime/RelayRenewProbe.swift
import Foundation

/// The data-path proof frame of `relay-renew/1`: a fixed 59-byte control frame
/// on the EXISTING text DataChannel that consumes no AEAD sequence and changes
/// no file or text semantics.
///
/// ## Why a dedicated frame and not the text codec
///
/// Routing this handshake through ordinary kind-9 content would be a hard
/// failure on all three clients whenever the conversation is not open — Web
/// `mixed-text-session.onData`'s `markFailed`, Android `TextLaneSession`'s
/// `CONTENT_BEFORE_ACTIVATION`, and here `LinkTextSession.admitFrame`'s
/// `failClosed` partition — and it would spend the strictly increasing text
/// `seq`. `0x0d` is outside every kind already in use (file lane 1-8, text 9,
/// fragmentation 10-11, pre-upload 12, and the `0xf8`-`0xff` lifecycle bytes),
/// and `link:§7.3` already requires a text-lane frame whose first byte is not
/// `0x09` to be SILENTLY IGNORED. That is what makes this frame safe to send to
/// a peer that has never heard of renewal.
///
/// ## Why HMAC is enough, and what it does not prove
///
/// The frame carries no secret — an epoch, a round, a random nonce. The
/// adversary that matters is a signalling-side DTLS man-in-the-middle, and it
/// cannot produce `resumeAuth`. Replay and reflection are closed by `from`/`to`,
/// `epoch`, `round`, the per-attempt random nonce, and the domain-separating
/// `kind` that makes a probe and its ack different strings.
///
/// **The tag authenticates key possession and freshness. It does NOT
/// authenticate the path.** The path evidence is the conjunction in
/// `RelayRenewController`: this side observed its selected local candidate
/// belongs to this epoch's ufrag generation, the peer only acks after its own
/// observation held, and the ack arrives after this side's observation held.
/// Nothing here may be described as proving a route.

/// The newly registered first byte. Structurally disjoint from every other kind
/// on either lane, which is what lets the demux claim it without inspecting
/// anything a peer could make ambiguous.
public let RELAY_RENEW_PROBE_KIND: UInt8 = 0x0d
public let RELAY_RENEW_PROBE_VERSION: UInt8 = 1
/// `[kind][ver][type][epoch u32BE][round u32BE][nonce 16][tag 32]`.
public let RENEW_PROBE_FRAME_BYTES = 59
public let RENEW_PROBE_NONCE_BYTES = 16
public let RENEW_PROBE_TAG_BYTES = 32

public enum RelayRenewProbeType: UInt8, Equatable, Sendable, CaseIterable {
    case probe = 1
    case ack = 2

    /// The domain separator in the signed payload. A probe and its ack are
    /// different strings, so an observed probe cannot be reflected back as its
    /// own acknowledgement.
    var payloadKind: String {
        switch self {
        case .probe: return "link-renew-probe"
        case .ack: return "link-renew-ack"
        }
    }
}

/// One parsed control frame. Structural only: nothing here has been verified,
/// and the tag is carried raw so the caller can decide — under its own budget,
/// and only after the cheap checks — whether to spend an HMAC on it.
public struct RelayRenewProbeFrame: Equatable, Sendable {
    public let type: RelayRenewProbeType
    public let epoch: UInt32
    public let round: UInt32
    public let nonce: [UInt8]
    public let tag: [UInt8]

    public init(type: RelayRenewProbeType,
                epoch: UInt32,
                round: UInt32,
                nonce: [UInt8],
                tag: [UInt8]) {
        self.type = type
        self.epoch = epoch
        self.round = round
        self.nonce = nonce
        self.tag = tag
    }

    /// The nonce as the signed payload spells it: standard RFC 4648 base64 with
    /// padding.
    public var nonceBase64: String { Data(nonce).base64EncodedString() }
}

/// The exact bytes a probe or ack tag covers.
///
/// Hand-serialised with a fixed key order for the reason `relayRenewPayload`
/// records. `from`/`to` are the established peer ids and are deliberately not
/// in the frame — a relay reflecting a frame at its sender verifies the
/// reversed tuple and fails.
public func relayRenewProbePayload(type: RelayRenewProbeType,
                                   from: String,
                                   to: String,
                                   epoch: UInt32,
                                   round: UInt32,
                                   nonce: [UInt8]) -> String {
    "{\"kind\":\"\(type.payloadKind)\",\"from\":\"\(escapeJSONStringValue(from))\""
        + ",\"to\":\"\(escapeJSONStringValue(to))\",\"epoch\":\(epoch),\"round\":\(round)"
        + ",\"nonce\":\"\(Data(nonce).base64EncodedString())\"}"
}

/// Serialise one control frame. `tag` must already be the raw 32-byte HMAC.
public func relayRenewProbeFrame(type: RelayRenewProbeType,
                                 epoch: UInt32,
                                 round: UInt32,
                                 nonce: [UInt8],
                                 tag: [UInt8]) -> [UInt8]? {
    guard nonce.count == RENEW_PROBE_NONCE_BYTES, tag.count == RENEW_PROBE_TAG_BYTES else {
        return nil
    }
    var out: [UInt8] = []
    out.reserveCapacity(RENEW_PROBE_FRAME_BYTES)
    out.append(RELAY_RENEW_PROBE_KIND)
    out.append(RELAY_RENEW_PROBE_VERSION)
    out.append(type.rawValue)
    out.append(contentsOf: bigEndianBytes(epoch))
    out.append(contentsOf: bigEndianBytes(round))
    out.append(contentsOf: nonce)
    out.append(contentsOf: tag)
    return out
}

private func bigEndianBytes(_ value: UInt32) -> [UInt8] {
    [UInt8(truncatingIfNeeded: value >> 24),
     UInt8(truncatingIfNeeded: value >> 16),
     UInt8(truncatingIfNeeded: value >> 8),
     UInt8(truncatingIfNeeded: value)]
}

/// Whether one text-lane frame belongs to renewal AT ALL.
///
/// Deliberately the kind byte alone, and deliberately separate from the parse.
/// This is the demux's question, and its answer must not depend on anything a
/// peer can make ambiguous: a frame whose first byte is `0x0d` is not a text
/// frame (`isLinkTextFrame` requires `0x09`), is not a lifecycle control, and
/// is not routable to anything else, so claiming it here is what keeps a
/// malformed renewal frame away from the text session's rate budget and
/// activity clock as well. Whether it is WELL-FORMED is
/// `parsedRelayRenewProbeFrame`'s question, and a malformed one is dropped in
/// silence rather than handed on.
public func isRelayRenewProbeFrame(_ frame: [UInt8]) -> Bool {
    frame.first == RELAY_RENEW_PROBE_KIND
}

/// Parse one control frame, structurally, before any HMAC.
///
/// The verification order spec §6.4 fixes starts here: length, kind byte,
/// version, type. Epoch/round matching, the per-epoch budget and the HMAC are
/// all the controller's, in that order, and every one of them is cheaper than
/// the one after it.
public func parsedRelayRenewProbeFrame(_ frame: [UInt8]) -> RelayRenewProbeFrame? {
    guard frame.count == RENEW_PROBE_FRAME_BYTES else { return nil }
    guard frame[0] == RELAY_RENEW_PROBE_KIND else { return nil }
    guard frame[1] == RELAY_RENEW_PROBE_VERSION else { return nil }
    guard let type = RelayRenewProbeType(rawValue: frame[2]) else { return nil }
    let epoch = UInt32(frame[3]) << 24 | UInt32(frame[4]) << 16
        | UInt32(frame[5]) << 8 | UInt32(frame[6])
    let round = UInt32(frame[7]) << 24 | UInt32(frame[8]) << 16
        | UInt32(frame[9]) << 8 | UInt32(frame[10])
    return RelayRenewProbeFrame(type: type,
                                epoch: epoch,
                                round: round,
                                nonce: Array(frame[11..<27]),
                                tag: Array(frame[27..<59]))
}

// MARK: - the front demux

/// What the text lane's front demux decided about one inbound frame.
public enum RelayRenewDemuxDisposition: Equatable, Sendable {
    /// Not renewal's. Hand it on to the text session unchanged.
    case pass
    /// Claimed and consumed. It must NOT also reach the text session — not its
    /// activity clock, not its rate budget, and under no circumstances its AEAD
    /// receiver.
    case consume
}

/// The single-slot front demux installed at the lanes' atomic attach point.
///
/// ## Why a true front demux and not an observer
///
/// An additive observer cannot STOP a frame from also reaching the session, and
/// stopping it is the entire requirement: a renewal frame that reached the text
/// session would (on the ports that keep one there) reset a ten-minute idle
/// timer this feature is explicitly forbidden from touching, and would spend an
/// inbound rate token a burst could then fail the lane on.
///
/// ## Lifetime is the TRANSPORT's, not the conversation's
///
/// It is installed where `LinkLaneOwner` builds the file driver's `onTextFrame`
/// route, which is the same atomic point the two lanes are attached at, so
/// frames replayed from `link:§2.2` pre-attachment capture — including across a
/// `link:§8` rebuild — go through exactly this routing. With no handler
/// installed a claimed frame is DROPPED and never replayed: a control frame
/// nobody is listening for is not a message, and holding it would be a buffer a
/// peer controls.
///
/// Thread-safe: WebRTC delivers on its own threads, and the handler is
/// installed and cleared from the controller's lifetime.
public final class RelayRenewProbeDemux: @unchecked Sendable {
    private let lock = NSLock()
    private var handler: ((RelayRenewProbeFrame) -> Void)?

    public init() {}

    /// Install or clear the renewal handler. Idempotent and last-writer-wins:
    /// one controller owns one transport's renewal at a time, and a controller
    /// closing after its successor installed must not be able to erase the
    /// successor — so callers clear through `remove(_:)`, not through `nil`.
    public func install(_ handler: @escaping (RelayRenewProbeFrame) -> Void) {
        lock.lock(); defer { lock.unlock() }
        self.handler = handler
    }

    /// Drop the handler. A frame claimed after this is consumed and discarded,
    /// which is the documented no-listener behaviour — never forwarded to the
    /// text session as a second chance.
    public func clear() {
        lock.lock(); defer { lock.unlock() }
        handler = nil
    }

    /// Route one text-lane frame. Called on WebRTC's delivery thread, ahead of
    /// `LinkTextDriver.admitTextFrame`.
    public func route(_ frame: [UInt8]) -> RelayRenewDemuxDisposition {
        guard isRelayRenewProbeFrame(frame) else { return .pass }
        // Claimed on the kind byte. Everything below may fail; none of it may
        // turn this back into a frame the text session sees.
        lock.lock()
        let handler = self.handler
        lock.unlock()
        guard let handler, let parsed = parsedRelayRenewProbeFrame(frame) else { return .consume }
        handler(parsed)
        return .consume
    }
}
